### ISSUE TITLE
Issue in training dataset with arrival times

### ISSUE THREAD

**Issue by pauldiac** [hobbit-project/DEBS-GC-2018#24]
Hello,

 We noticed a problem that occurs repeatedly in the training dataset. The arrival time in some ship trips seems to be the time reported by the ship when entering the radius of the first port encountered, and not the time of reaching the reported arrival port. Examples:
 
 * For ship with id 0x00e225cd5b1926c6b0eb4683692bc2f0ba9a3a2b on a trip from DILISKELESI to ALEXANDRIA the reported arrival time is "26-04-15 20:31". At this time the ship is at (Lon, Lat) = (29.3985, 40.74567), which is the first reported point within the radius of the TUZLA port (not ALEXANDRIA).
 
 * For ship with id 0x019d2386bff02a0aa96e825e56668dc359e3a987 on a trip from DILISKELESI to GEMLIK the reported arrival time is "12-05-15 4:05". At this time the ship is at (Lon, Lat) = (29.39127, 40.74445), which is the first reported point within the radius of the TUZLA port (not GEMLIK).
 
 * For ship with id 0x057e6e20d32a9e595926f5f74efce899bad96d97 on a trip from PALMA DE MALLORCA to GENOVA the reported arrival time is "14-05-15 6:01". At this time the ship is at (Lon, Lat) = (7.361688, 43.64136), which is the first reported point within the radius of the MONACO port (not GENOVA).

 Should we consider the arrival port, the first port within whose radius the ship entered? or the arrival time is wrong and should be corrected?
 
 
 We have also noticed some other strange cases in the dataset, such as ship with id 0x023db8bbb80aa3144444cefe369f4e4edfedd70f on a reported trip from PALMA DE MALLORCA to VALENCIA travels this route back and forth several times, but in the dataset this is considered as a single travel. To clarify the issue, the ship arrives to VALENCIA as reported in the dataset at "14-05-15 4:51" (the first point reported within VALENCIA radius). Afterwards looking at the coordinates emitted by the ship, this returns to PALMA DE MALLORCA, but the reported arrival time does not change. Moreover, the departure port and arrival port remains the same (PALMA DE MALLORCA to VALENCIA although the ship moves back from VALENCIA to PALMA DE MALLORCA). Is this behavior intended?

Thank you,

**Comment by smirnp** [comment]
Hi! 

> Should we consider the arrival port, the first port within whose radius the ship entered? or the arrival time is wrong and should be corrected?

That's a valid point. The correct behavior is to consider the timestamp and name of the last intersecting port. 
I have checked the training data and find that about 10% of trips labeled incorrectly (the timestamp of first intersecting port was taken as label, while the name of the port was taken correctly).

 The updated (fixed) training dataset as well as the announcement of additional training dataset only for participants, who uploaded working system to the platform will be announced soon.

> We have also noticed some other strange cases in the dataset, such as ship with id 0x023db8bbb80aa3144444cefe369f4e4edfedd70f on a reported trip from PALMA DE MALLORCA to VALENCIA travels this route back and forth several times, but in the dataset this is considered as a single travel. To clarify the issue, the ship arrives to VALENCIA as reported in the dataset at "14-05-15 4:51" (the first point reported within VALENCIA radius). Afterwards looking at the coordinates emitted by the ship, this returns to PALMA DE MALLORCA, but the reported arrival time does not change. Moreover, the departure port and arrival port remains the same (PALMA DE MALLORCA to VALENCIA although the ship moves back from VALENCIA to PALMA DE MALLORCA). Is this behavior intended?

We agree, what such trips exist, but decided to leave them unsplitted, to save an original quality of data. Such deviating trips will be excluded from the rank results assessment/evaluation.

**Comment by eonica** [comment]
Dear Pavel,

Thank you for the response. Could you please clarify again what port should
we consider as arrival point?
You mentioned that the correct behavior should be considering the timestamp
and the name of the intersecting port - to what intersecting port do you
refer? the first or the last?

Also, the deviating trips you mentioned that will be excluded from the rank
evaluation will include just trips composed from multiple trips that were
not split between two ports - back and forth, like the one in the given
example. Or also trips composed from multiple trips between various other
ports? e.g., a trip reported as being from departure port A to destination
port C, but being actually composed from two trips from A to B and B to C
that were not split. This detail can be significant for the training phase
of the solution.

Best regards,
Emanuel

**Comment by smirnp** [comment]
Dear all,

Today an updated (fixed) training dataset will be published. 
Trips with multiple ports (about 10% of all trips) will be splitted and labeled separately. 
Finally all the trips in data (the training and the test one) will have only two ports: departure and arrival.

**Comment by acip** [comment]
Hi @smirnp ,

Do you have an update on the updated dataset?

Thanks!

**Comment by smirnp** [comment]
Hi! Here is the [updated file](https://hobbitdata.informatik.uni-leipzig.de/DEBS_GC_2018/DEBS%202018_training_fixed_5.ZIP), but I'm still working on that.
The file contains about 15% on unlabeled tuples, which are:
a) tuples, when ship already entered arrival port radius, but continues send tuples for some time
b) trips were finished by (a) but continue to send tuples under the same trip_Id for next weeks and months. 

Ships with parkings during the trips (such as 0x6d679_15-03-15 20:xx) still have incorrect labels. 
Currently I'm working on algorithm to split trips by: 
a) arriving to some port and then departure
b) continuous stopping during the trip

Hopefully today or tomorrow I will solve that problems

**Comment by smirnp** [comment]
Please find the [updated file](https://hobbitdata.informatik.uni-leipzig.de/DEBS_GC_2018/DEBS%202018_training_fixed_6.ZIP). Trips were splitted by long parkings times,  parking tuples were removed. Labelling was done by the first arrival port. Short trips started to look better. Still there is a problem with tuples, where ships did not updated their departure ports from the previous trips, but actually leaved port. Such trips will be splitted and labeled during the next week.

**Comment by nduy** [comment]
Thank you for your effort! As you added TRIP_ID field to the data, is that correct that the evaluation is now in term of TRIP_ID instead of SHIP_ID as it was before? Indeed, averageEarlynessRate and meanErrorMin is calculated between trips, is that correct @smirnp?
And if that is correct, the implementation of evaluation module may need some necessary changes as well, as far as I understand.

**Comment by smirnp** [comment]
Hi, @nduy !

> averageEarlynessRate and meanErrorMin is calculated between trips

yes. Such a way is considered from the beginning. The `EalynessRate` is calculated per trip, and then the average is calculated as sum(rates)/N trips. The `meanErrorMin` - is taken per trip, then sum(errors)/N trips

> And if that is correct, the implementation of evaluation module may need some necessary changes as well, as far as I understand. 

No, the evaluation module already calculates results as described. The small dataset in the online benchmark is needs to be updated according to the latest changes in data labeling algorithms.   

In fact the `TripID` is a service field (such as `arrival_calc` and `arrival_port_calc`), which is necessary to obtain correct ground truth results for evaluation. It will be not send in  data to the benchmarked systems (please [see](https://github.com/hobbit-project/DEBS-GC-2018/blob/94b18c78ce39c9df2e9f300d1e238186f3f02b0e/src/main/java/org/hobbit/debs_2018_gc_samples/Benchmark/TaskGenerator.java#L335) that only `raw` data will be send)  - the same way as the data is coming from the ships into real maritime systems.  

Even without the `TripID` tuples are considered to be grouped into trips and an evaluation to be performed over trips - not over all tuples of the ship as a single trip. Initially tuples were considered to be grouped into trips by changes of the declared DeparturePortName and preserving timestamp order (e.g. all consequent tuples of a ship have the same tripId=`DeparturePortName+<tripIndex>` until DeparturePortName changes - e.g. after ship arrived to port and updated it's a DeparturePortName for a new trip). 

But such grouping was is not correct because: 
a) Ships do not always update their departure port names (according to coordinates ship arrive to DeparturePort and then depart from it still having the same DeparturePortName); 
b) Having the same DeparturePortName ships continue to send tuples being parked somethere for weeks or months and then continue going to DeparturePortName or any other place. It means that DeparturePortName is a manual entering field and only sensor measurements (coordinates, speed, etc) can reflect the real behavior/trips. 

So it was decided to split the tuples by 
a) crossing the radius of some port (arrival); 
b) long parking times (remove them at all); 
c) leaving the raduis of some port after arriving to it;  
and to introduce the `TripID=shipId_TimeStampOfAFirstTupleInAGroup`  to consider the grouped tuples as separate trips. 

(a) and (b) was already done, (c) is in progress (such tuples are not labeled now).

**Comment by nduy** [comment]
Thank you for your thoughtful explanation. Hope the (c) will be completed soon ;)

**Comment by tsung-wei-huang** [comment]
@smirnp  Thanks for the reply. Based on this update, data during online evaluation will be sent on a trip basis, correct? Suppose there are 100 trips, each with different length. The online benchmark will have 100 queues line up. Each queue generates a tuple and the next tuple will not be delivered until receiving our prediction. **Inter-tuples arrive concurrently while intra-tuples arrive sequentially**. Please confirm if our understanding is correct. Previously we thought it was per ship basis.

**Comment by smirnp** [comment]
@twhuang-uiuc , potentially trips of single ship might be served in parallel, but since it creates opportunities for cheating (using information about further trips of a ship) and becomes not realistic anymore (in real systems ships are sending their tuples in realtime manner and can belong only to one trip at the moment) such a variant was avoided. 
Currently trips of a single ship are served in sequentially (tuple by tuple), but all the ships are served in parallel. 
Currently all ships serving via a single queue filled with a tuples with different values.

**Comment by tsung-wei-huang** [comment]
I see. Thanks.

**Comment by clin99** [comment]
@smirnp  

1. Based on the explanation, can we assume during online evaluation our System will only receive a tuple per ship at the same time?

2. Can we also assume during online evaluation successive trips of a ship will **never** have the same departure port?    For example, a ship might have successive trips like: _port A -> port B (trip 1),  port B -> port C (trip 2)_ but not:  _port A -> port B (trip 1), port A -> port C (trip 2)_

Thanks.

**Comment by smirnp** [comment]
1. Yes. Only one tuple of a ship in the moment.
2. No. Some trips are missed, so it might be A->B, then A->C and so on.

**Comment by bdnf** [comment]
Hi, @smirnp !

I've noticed that the latest dataset still contains a lot of missing data.
Could you please make an update when the corrected dataset will be available?

Many thanks in advance!

**Comment by smirnp** [comment]
Hi! Yes, currently I'm working on improvements. Hopefully and update will be in the next few days.

**Comment by bdnf** [comment]
Thank you, Pavel

Will be the structure of dataset preserved or should we expect some extension of the deadline to be able to deal with the updated data?

**Comment by nduy** [comment]
Many thanks for your work, Pavel. I have just notice cases that ship parked at the middle of sea for days but apparently, trips were not splitted. For instance, trip 0xc35c9_03-04-15 16:xx parked from `06-04-15 8:03` to `07-04-15 2:19` (almost 1 day); and `07-04-15 2:43` to `09-04-15 17:22` (more than 2 days). Are you planning to deal with this as well, or you left these cases that way intentionally?

**Comment by smirnp** [comment]
@bdnf , the data itself and its structure remains the same. 
The labels, which will be used as ground truth for the evaluation were improved. You may update images of the registered system until the final execution (end of May).

**Comment by smirnp** [comment]
@nduy , in last version of labels (v7) this ship is excluded. 
Small parkings, which do not affect much for the whole trip are left.

**Comment by Abder88** [comment]
Which training datasets we must use?
Thanks in advance!

**Comment by smirnp** [comment]
Hi, @Abder88!
You may use both datasets (first and second), they are not overlapping.
Unfortunately registration of new systems/paper submissions for the challenge was closed by the 7th of May.

**Comment by Abder88** [comment]
Hi @smirnp,
We have a paper accepted but we are using the first training dataset defined from the beginning of the challenge!

**Comment by smirnp** [comment]
@Abder88 , please request an access for the second dataset as descibed under the News section of this repo.
